lib/uktest: Add README.md file (#555)
unikraft/unikraft: lib/uktest/README.md

> # Unikraft’s Testing Framework.

Reviewer: ```suggestion
# uktest: Unikraft’s Testing Framework
```

> 

Reviewer: I would add a short paragraph stating what uktest is (even if it's quite obvious), what its purpose is and how it should be used.

## Overview

In `uktest`, tests are organised hierarchically powered by the the lowest common denominator: the assertion.
This organisation is inspired by [KUnit](https://kunit.dev/), the Linux Kernel's in-house testing system.
For licensing reasons the Unikraft project cannot use this source code.
However, by inspiration, we can organise the `uktest` library following a similar pattern:

> 1. The assertion: it repersents the lowest common denominator of a test: some boolean operation which, when true, a single test passes.

Reviewer: I would use a simple list with assertion, test case and test suite ad their one sentence definition.

Then, I would add a subsection (`###`) detailing each.

1. The test case: often, assertions are not alone in their means to check the legitimacy of operation in some function.
We find that a "case" is best way to organise a group of assertions in which one particular function of some system is under-going testing.
A case is independent of other cases, but related in the same sub-system.
For this reason we register them together in the same test suite.

1. The test suite: represents a group of test cases.
This is the final and upper-most heirarchical repesentation of tests and their groupings.
With assertions grouped into test cases and test cases grouped into a test suite, we end this organisation in a fashion which allows us to follow a common design pattern within Unikraft: the registration model.
The syntax follows similar to other registation models within Unikraft, e.g. `ukbus`.
However, `uktest`'s registation model is more powerful.

## Creating Tests

To register a test suite with `uktest`, we simply invoke `uk_testsuite_register` with a unique symbol name.
This symbol is used along with test cases in order to create the references to one-another.

> Each test case has only two input parameters: a reference to the suite is part, as well as a canonical name for the case itself.

Reviewer: ```suggestion
Each test case has two input parameters: a reference to the suite is part, as well as a canonical name for the case itself.
```

Generally, the following pattern is used for test suites:

```
$LIBNAME_$TESTSUITENAME_testsuite
```

An the following for test cases:

```
$LIBNAME_test_$TESTCASENAME
```

> To create a case, simply invoke the `UK_TESTCASE` macro with the two parameters describe previously, and use in the context of a function, for example:

Reviewer: ```suggestion
To create a case, invoke the `UK_TESTCASE` macro with the two parameters describe previously, and use in the context of a function, for example:
```

```c
UK_TESTCASE(uktest_mycase_testsuite, uktest_test_case)
{
        int x = 1;
        UK_TEST_EXPECT(x > 0);
}
```

Finally, to register the case with a suite (see next section), call one of the possible registration functions:

```c
uk_testsuite_register(uktest_mycase_testsuite, NULL);
```

> The above snippet can be organised within a library in a number of ways such as in-line or as an individual file representing the suite.

Reviewer: ```suggestion
The above snippet can be organised within a library in a number of ways, such as in-line or as an individual file representing the suite.
```

There are a number of test suite registration handles which are elaborated on in next section.
It should be noted that multiple test suites can exist within a library in order to test multiple features or components of said library.

In order to achieve consistency in the use of `uktest` across the Unikraft code base, the following recommendation is made regarding the registration of test suites:

1. A single test suite should be organised into its own file, prefixed with `test_` (e.g. `test_feature.c`).

> All tests suites of some library within Unikraft should be stored within a new folder located at the root of the library named `tests/`.

Reviewer: ```suggestion
   All tests suites of some library within Unikraft should be stored within a new folder located at the root of the library named `tests/`.
```

# uktest: Unikraft’s Testing Framework

`uktest` is Unikraft's internal testing framework.

> It provides a flexible API for performing unit testing internal libraries, methods and more in order to safeguard changes which affect existing functionality.

Reviewer 2: Apologies, I see we're missing an `of` in the sentence:

```diff
- It provides a flexible API for performing unit testing internal libraries, methods and more in order to safeguard changes which affect existing functionality.
+ It provides a flexible API for performing unit testing of internal libraries, methods and more in order to safeguard changes which affect existing functionality.
```

Author: Ah, I missed that.

# uktest: Unikraft’s Testing Framework

`uktest` is Unikraft's testing framework, has been inspired by [KUnit](https://kunit.dev/) and provides a flexible testing API.
In order to use the API, you need to include `uk/test.h`.

> The exposed functions are detailed below and on [Unikraft's website](https://unikraft.org/).

Reviewer: ```suggestion
The exposed functions are detailed below and on [the Unikraft documentation](https://unikraft.org/docs/develop/writing-tests/).
```

> The exposed functions are detailed below and on [the Unikraft's documentation](https://unikraft.org/docs/develop/writing-tests/).

Reviewer 2: ```diff
- `uktest` is Unikraft's testing framework, has been inspired by [KUnit](https://kunit.dev/) and provides a flexible testing API.
- In order to use the API, you need to include `uk/test.h`.
- The exposed functions are detailed below and on [the Unikraft's documentation](https://unikraft.org/docs/develop/writing-tests/).
+ `uktest` is Unikraft's internal testing framework.
+ It provides a flexible API for performing unit testing internal libraries, methods and more in order to safe-guard changes which affect existing functionality.
```

## Overview

In `uktest`, tests are organised hierarchically powered by the the lowest common denominator: the assertion.
This organisation is inspired by [KUnit](https://kunit.dev/), the Linux Kernel's in-house testing system.
For licensing reasons the Unikraft project cannot use this source code.
However, by inspiration, we can organise the `uktest` library following a similar pattern:

1. The assertion: it repersents the lowest common denominator of a test: some boolean operation which, when true, a single test passes.

1. The test case: it represents a group of assertions in which one particular function or some system is under-going testing.

1. The test suite: represents a group of test cases.

### Assertions

Assertions are often used in-line and their usage should be no different to the traditional use of the `ASSERT` macro.

> In `uktest`, we introduce a new definition: `UK_TEST_EXPECT` which has one parameters: the same boolean opeation which is true-to-form of the traditional `ASSERT` macro.

Reviewer: ```suggestion
In `uktest`, we introduce a new definition: `UK_TEST_EXPECT` which has one parameter: the same boolean operation which is true-to-form of the traditional `ASSERT` macro.
```

With `uktest`, however, the macro is intelligently placed in context within a case (see the next point).
Additional text or descriptive explanation of the text can also be provided with the auxiliary and similar macro `UK_TEST_ASSERTF`.

### Test Cases

The test case: often, assertions are not alone in their means to check the legitimacy of operation in some function.
We find that a "case" is best way to organise a group of assertions in which one particular function of some system is under-going testing.
A case is independent of other cases, but related in the same sub-system.
For this reason we register them together in the same test suite.

> ### Test Suits

Reviewer: ```suggestion
### Test Suites
```

The test suite: represents a group of test cases.
This is the final and upper-most heirarchical repesentation of tests and their groupings.
With assertions grouped into test cases and test cases grouped into a test suite, we end this organisation in a fashion which allows us to follow a common design pattern within Unikraft: the registration model.

> The syntax follows similar to other registation models within Unikraft, e.g. `ukbus`.

Reviewer: ```suggestion
The syntax is similar to other registration models within Unikraft, e.g. `ukbus`.
```

However, `uktest`'s registation model is more powerful.

## Creating Tests

To register a test suite with `uktest`, we simply invoke `uk_testsuite_register` with a unique symbol name.

> This symbol is used along with test cases in order to create the references to one-another.

Reviewer: ```suggestion
This symbol is used along with test cases in order to create references to one-another.
```

Each test case has two input parameters: a reference to the suite is part, as well as a canonical name for the case itself.
Generally, the following pattern is used for test suites:

```
$LIBNAME_$TESTSUITENAME_testsuite
```

An the following for test cases:

```
$LIBNAME_test_$TESTCASENAME
```

> To create a case, invoke the `UK_TESTCASE` macro with the two parameters describe previously, and use in the context of a function, for example:

Reviewer: ```suggestion
To create a case, invoke the `UK_TESTCASE` macro with the two parameters describe previously, and use it in the context of a function, for example:
```

```c
uk_testsuite_register(uktest_mycase_testsuite, NULL);
```

The above snippet can be organised within a library in a number of ways, such as in-line or as an individual file representing the suite.
There are a number of test suite registration handles which are elaborated on in next section.
It should be noted that multiple test suites can exist within a library in order to test multiple features or components of said library.

In order to achieve consistency in the use of `uktest` across the Unikraft code base, the following recommendation is made regarding the registration of test suites:

1. A single test suite should be organised into its own file, prefixed with `test_` (e.g. `test_feature.c`).
   All tests suites of some library within Unikraft should be stored within a new folder located at the root of the library named `tests/`.

1. All tests suites should have a corresponding KConfig option, prefixed with the library name and then the word "TEST" (e.g. `LIBNAME_TEST_`).

1. Every library implementing one or more suite of tests must have a new menuconfig housing all test suite options under the name `LIBNAME_TEST`.

> This menuconfig option must invoke all the suites if `LIBUKTEST_ALL` is set to true.

Reviewer: ```suggestion
   This menuconfig option must invoke all the suites if `LIBUKTEST_ALL` is set to true.
```
